module_utils/certificate/providers/certmonger.py

```diff
@@ -0,0 +1,142 @@
+        #   only difference from an actual ini-file is that it does not
+        #   make use of sections. Since ConfigParser can't parse files
+        #   without section we add a single section using an in-memory
+        #   copy of the file.
+        config = ConfigParser()
+        config_string_file = StringIO()
+        config_string_file.write("[certmonger]\n")
+        with open(self._request_file_path, "r") as config_file:
+            config_string_file.write(config_file.read())
+        config_string_file.seek(0)
+
+        config.readfp(config_string_file)
+        return dict(config["certmonger"])
+
+    def get_existing_csr_pem_data(self):
+        """Read CSR PEM data from certmonger metadata."""
+        return self._certmonger_metadata.get("csr").encode("utf-8")
+
+    def _get_certmonger_ca_from_params(self):
+        ca = self.module.params.get("ca")
+        if ca == "self-sign":
+            ca = "local"
+        return ca
+
+    def _get_certmonger_ca_for_existing_cert(self):
+        return self._certmonger_metadata.get("ca_name")
+
+    @property
+    def cert_needs_update(self):
+        """Check if the existing_certificate needs update.
+
+        Since the provider base class doesn't know about CA this method
+        needs to add the verification for CA change.
+        """
+        needs_update = super(
+            CertificateRequestCertmongerProvider, self
+        ).cert_needs_update
+        ca_from_params = self._get_certmonger_ca_from_params()
+        ca_from_existing_cert = self._get_certmonger_ca_for_existing_cert()
+        if needs_update or ca_from_params != ca_from_existing_cert:
+            return True
+
+        return False
+
+    @property
+    def exists_in_certmonger(self):
+        """Check if certificate is tracked by certmonger."""
+        if self._certmonger_metadata:
+            return True
+        return False
+
+    def request_certificate(self):
+        """Issue or update a certificate using certmonger."""
+        getcert_bin = self.module.get_bin_path("getcert", required=True)
+        command = [getcert_bin]
+
+        if self.exists_in_certmonger:
+            command += ["resubmit"]
+        else:
+            command += ["request"]
+
+        # Set common name
+        command += ["-N", self.csr.common_name]
+
+        # Set CA
+        command += ["-c", self._get_certmonger_ca_from_params()]
+
+        # Wait for cert
+        command += ["-w"]
+
+        # Set certificate locations
+        if not self.exists_in_certmonger:
+            command += ["-k", self.certificate_key_path]
+        command += ["-f", self.certificate_file_path]
+
+        # Set Domains
+        for dns in self.csr.dns:
+            command += ["-D", dns]
+
+        if not self.exists_in_certmonger:
+            # Don't attempt to renew when near to expiration
+            command += ["-R"]
+
+        self._run_command(command, check_rc=True)
```

Comment:
So this actually requests the cert - I don't think we need a handler, unless we need the handler to check to see if the cert request was successful?  Does `-w` here mean that the cert request task will wait until the certificate and key are issued and placed on disk?  If a cert request requires human interaction e.g. a security officer approving the cert request, might that take hours or so?

Reply:
We'll have an option for not waiting. This is just the base structure so I can start work to add the actual supported parameters.

Reply:
ok - something to consider when developing the handlers